onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -1012,6 +1021,12 @@ TensorrtExecutionProvider::GetCapability(const GraphViewer& graph,
   return result;
 }
 
+std::unique_lock<OrtMutex> TensorrtExecutionProvider::GetEngineBuildLock() const {
+  static OrtMutex singleton;
+  std::unique_lock<OrtMutex> lock = force_sequential_engine_build_ ? std::unique_lock<OrtMutex>(singleton) : std::unique_lock<OrtMutex>();
```

Comment:
this lock is only needed for the same model running on multiple GPUs. If different models running on multiple GPUs, do we still need to enable sequential_engine_build?

Reply:
Actually the lock needs to be global, i.e. shared across all models/providers/sessions. I did test it locally by loading a different model (each with different path) at each GPU (on my 8-gpu machine), and unfortunately it still runs into access violations. The same issue does not repro when force_sequential_engine_build_  is enabled.

Semantically, when force_sequential_engine_build_  is set, it means there could be only a single TRT engine being built at the same time -- regardless of whether from different models or even different subgraphs of the same model.

Reply:
Thanks. That's interesting. Do you see any perf degradation before and after the lock is added? Also it would be better to run real queries (with different input shapes) and check if there is any issue.

Reply:
per TRT docs, https://docs.nvidia.com/deeplearning/tensorrt/best-practices/index.html#thread-safety
"The TensorRT builder may only be used by one thread at a time. If you need to run multiple builds simultaneously, you will need to create multiple builders."
I wonder why multiple sessions/ExecutionProviders with lock (non-global) per Provider didn't work.
We may need to clarify with Nvidia/TensorRT devs.

Reply:
Yes. There is one builder for each session and it should not need global lock. That's something I will discuss with Nvidia.